spec/service/api/v2/form_document_sync_service_spec.rb

```diff
@@ -0,0 +1,133 @@
+    end
+  end
+
+  describe "#sync_live_with_draft_form" do
+    let(:form) { create(:form, :live) }
+
+    it "updates or creates live and draft form documents and deletes archived document" do
+      expect(service).to receive(:update_or_create_form_document).with(form, :live, anything)
+      expect(service).to receive(:update_or_create_form_document).with(form, :draft, anything)
+      expect(service).to receive(:delete_form_documents).with(form, [:archived])
+      service.sync_live_with_draft_form(form)
+    end
+  end
+
+  describe "#sync_draft_form" do
+    it "updates or creates draft form document and deletes live and archived documents" do
+      expect(service).to receive(:update_or_create_form_document).with(form, :draft, anything)
+      expect(service).to receive(:delete_form_documents).with(form, %i[live archived])
+      service.sync_draft_form(form)
+    end
+  end
+
+  describe "#sync_archived_form" do
+    it "updates or creates archived form document and deletes live and draft documents" do
+      expect(service).to receive(:update_or_create_form_document).with(form, :archived, anything)
+      expect(service).to receive(:delete_form_documents).with(form, %i[live draft])
+      service.sync_archived_form(form)
+    end
+  end
+
+  describe "#sync_archived_with_draft_form" do
+    it "updates or creates archived and draft form documents and deletes live document" do
+      expect(service).to receive(:update_or_create_form_document).with(form, :archived, anything)
+      expect(service).to receive(:update_or_create_form_document).with(form, :draft, anything)
+      expect(service).to receive(:delete_form_documents).with(form, [:live])
+      service.sync_archived_with_draft_form(form)
+    end
+  end
+
+  describe "#update_or_create_form_document" do
+    let(:form_document) { instance_double(Api::V2::FormDocument, save!: true) }
+
+    before do
+      allow(Api::V2::FormDocument).to receive(:find_or_initialize_by).and_return(form_document)
+    end
+
+    it "finds or initializes a form document and updates its content" do
+      expect(Api::V2::FormDocument).to receive(:find_or_initialize_by).with(form_id: form.id, tag: :live)
+      expect(form_document).to receive(:content=)
+      expect(form_document).to receive(:save!)
+      service.update_or_create_form_document(form, :live, {})
+    end
+  end
+
+  describe "#delete_form_documents" do
+    it "deletes form documents with specified tags" do
+      form_documents = instance_double(ActiveRecord::Relation)
+      allow(Api::V2::FormDocument).to receive(:where).with(form_id: form.id, tag: %i[draft archived]).and_return(form_documents)
+      allow(form_documents).to receive(:delete_all).and_return(true)
+
+      service.delete_form_documents(form, %i[draft archived])
+
+      expect(Api::V2::FormDocument).to have_received(:where).with(form_id: form.id, tag: %i[draft archived])
+      expect(form_documents).to have_received(:delete_all)
```

Comment:
I think we could rewrite this test to use the database instead of mocking the model:

```suggestion
      create(:api_v2_form_document, form_id: form.id, tag: :draft)
      create(:api_v2_form_document, form_id: form.id, tag: :archived)
      create(:api_v2_form_document, form_id: form.id, tag: :live)
      
      expect {
          service.delete_form_documents(form, %i[draft archived])
      }.to change(Api::V2::FormDocuments, :count).from(3).to(0)
```

That way we're testing the behaviour rather than the implementation.

Reply:
nice idea, I've updated it